src/QMCDrivers/WFOpt/LinearMethod.cpp

```diff
@@ -0,0 +1,171 @@
+      mappedEigenvalues[i].second = i;
+    }
+  }
+  std::sort(mappedEigenvalues.begin(), mappedEigenvalues.end());
+  for (int i = 0; i < Nl; i++)
+    ev[i] = eigenT(mappedEigenvalues[0].second, i) / eigenT(mappedEigenvalues[0].second, 0);
+  return mappedEigenvalues[0].first;
+}
+
+LinearMethod::Real LinearMethod::getLowestEigenvector(Matrix<Real>& A, std::vector<Real>& ev) const
+{
+  int Nl(ev.size());
+  //   Getting the optimal worksize
+  Real zerozero = A(0, 0);
+  char jl('N');
+  char jr('V');
+  std::vector<Real> alphar(Nl), alphai(Nl), beta(Nl);
+  Matrix<Real> eigenT(Nl, Nl);
+  Matrix<Real> eigenD(Nl, Nl);
+  int info;
+  int lwork(-1);
+  std::vector<Real> work(1);
+  LAPACK::geev(&jl, &jr, &Nl, A.data(), &Nl, &alphar[0], &alphai[0], eigenD.data(), &Nl, eigenT.data(), &Nl, &work[0],
+               &lwork, &info);
+  lwork = int(work[0]);
+  work.resize(lwork);
+
+  LAPACK::geev(&jl, &jr, &Nl, A.data(), &Nl, &alphar[0], &alphai[0], eigenD.data(), &Nl, eigenT.data(), &Nl, &work[0],
+               &lwork, &info);
+  if (info != 0)
+  {
+    APP_ABORT("Invalid Matrix Diagonalization Function!");
+  }
+  std::vector<std::pair<Real, int>> mappedEigenvalues(Nl);
+  for (int i = 0; i < Nl; i++)
+  {
+    Real evi(alphar[i]);
+    if ((evi < zerozero) && (evi > (zerozero - 1e2)))
+    {
+      mappedEigenvalues[i].first  = (evi - zerozero + 2.0) * (evi - zerozero + 2.0);
+      mappedEigenvalues[i].second = i;
+    }
+    else
+    {
+      mappedEigenvalues[i].first  = std::numeric_limits<Real>::max();
+      mappedEigenvalues[i].second = i;
+    }
+  }
+  std::sort(mappedEigenvalues.begin(), mappedEigenvalues.end());
+  //         for (int i=0; i<4; i++) app_log()<<i<<": "<<alphar[mappedEigenvalues[i].second]<< std::endl;
+  for (int i = 0; i < Nl; i++)
+    ev[i] = eigenT(mappedEigenvalues[0].second, i) / eigenT(mappedEigenvalues[0].second, 0);
+  return alphar[mappedEigenvalues[0].second];
+  //     }
+}
+
+void LinearMethod::getNonLinearRange(int& first, int& last, const QMCCostFunctionBase& optTarget) const
+{
+  std::vector<int> types;
+  optTarget.getParameterTypes(types);
+  first = 0;
+  last  = types.size();
+  //assume all non-linear coeffs are together.
+  if (types[0] == optimize::LINEAR_P)
+  {
+    int i(0);
+    while (i < types.size())
+    {
+      if (types[i] == optimize::LINEAR_P)
+        first = i;
+      i++;
+    }
+    first++;
+  }
+  else
+  {
+    int i(types.size() - 1);
+    while (i >= 0)
+    {
+      if (types[i] == optimize::LINEAR_P)
+        last = i;
+      i--;
+    }
+  }
+  //     returns the number of non-linear parameters.
+  //    app_log()<<"line params: "<<first<<" "<<last<< std::endl;
+}
+
+LinearMethod::Real LinearMethod::getNonLinearRescale(std::vector<Real>& dP,
+                                                     Matrix<Real>& S,
+                                                     const QMCCostFunctionBase& optTarget) const
+{
```

Comment:
Passing a "heavy" or "large" class like QMCCostFunctionBase to a function like this caught my attention.  (Maybe a better description for QMCCostFunctionBase would be that it is higher up in a logical hierarchy of concerns.) 
An immediate question is maybe getNonLinearRange should be a member function of QMCCostFunctionBase? (I think the answer is: it should not be, because this function relates to the linear method, not the cost function)

It would be better to pass just what is needed - in this case only the vector of parameter types used in getNonLinearRange.
(Passing the entire class is reasonable as an intermediate step in refactoring the code).

Reply:
legacy CUDA cod removal has been taken out of this PR.